### PR TITLE
Use dependency tree for instruction of processing

### PR DESCRIPTION
**Pull request by dachengx**
Depends on https://github.com/AxFoundation/strax/pull/896

Close: https://github.com/XENONnT/outsource/issues/130
Close: https://github.com/XENONnT/outsource/issues/152
Close: https://github.com/XENONnT/outsource/issues/189
Close: https://github.com/XENONnT/outsource/issues/203

This a complex PR. But the main ideas are simple:

1. **Submit at most two jobs for a combination of `(run_id, detector)`.** If the `tpc` and `neutron_veto` are both available for of a `run_id`, outsource will submit at most four jobs. To accomplish this, this PR separates the dependency tree into "below per-chunk processing (lower job)" and "above per-chunk processing (upper job)". This is done in `RunConfig.get_needs_processed`. Later in the job submission, if the lower data_types are not processed, outsource will first submit a job to process lower data_types _per-chunk_, then combine them. And then submit a job to process the data_types in `include_data_types` in `XENON_CONFIG`.
2. **Use the dependency tree in strax(en) context to guide the job submission and processing.** This is done by adding multiple functions. Most of them are in `utils.py`: `get_possible_dependencies`, `get_to_save_data_types`, `get_rse`, and `per_chunk_storage_root_data_type`.

- `get_possible_dependencies` returns the data_types provided **along with** `PER_CHUNK_DATA_TYPES` and `root_data_types`(like `raw_records*`). So if `peaklets` is in `PER_CHUNK_DATA_TYPES`, `get_possible_dependencies` will return a set including `lone_hits`. Because sometimes maybe a plugin only depends on `lone_hits` but outsource only processes `peaklets`.
- `get_to_save_data_types` returns all the data_types that will be saved during processing.
- `get_rse` checks if a data_type is below `records`, if so we will save the data_type in `records_rse`, a similar rule applies to other data_types, please go to the details in that function.
- `per_chunk_storage_root_data_type` returns `raw_records` if it is below or at the same level as `PER_CHUNK_DATA_TYPES`. If not so, it will return `None`. This function tells you if the data_type is being processed per-chunk. For example, if data_type is `event_info`, it will return `None`, because the outsource will submit a job to (per-chunk-ly) process `peaklets`( and the data_type at the same level as `peaklets`). But if the data_type is `peaklets` or `hitlets_nv`, it will return `raw_records` or `raw_records_nv` so outsource will submit per-chunk processing and then a "combine" job.

### CODE REVIEW

File: outsource/meta.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
+# These data_types need per-chunk storage, so don't upload to rucio here!
+PER_CHUNK_DATA_TYPES = [
```
Comment (rsriya) [inline outsource/meta.py:2]: What about records?

Comment (dachengx) [reply outsource/meta.py:2]: `records` are not saved. So I disallow `records` being in `include_data_types`.

File: outsource/meta.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+]
+
+
+# Do a query to see if these data_types are present
+DETECTOR_DATA_TYPES = {
+    "tpc": {
+        "raw": "raw_records",
+        "to_process": [
+            "peaklets",
+            "event_ms_naive",
```
Comment (rsriya) [inline outsource/meta.py:16]: Why are event_info, peak_basics and event_area_per_channel removed?

File: outsource/meta.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
+# These data_types need per-chunk storage, so don't upload to rucio here!
+PER_CHUNK_DATA_TYPES = [
+    "peaklets",
```
Comment (rsriya) [inline outsource/meta.py:3]: Why is records removed?

File: outsource/config.py [hunk 4]
```diff
@@ -301,15 +123,37 @@ def get_requirements(self, rses):
...
+        """Get the data_type this one depends on.
+
+        The result will be either data_type in same level of PER_CHUNK_DATA_TYPES or the
+        root_data_types.
+
+        """
+        possible_dependencies = get_possible_dependencies(self.context)
+        # Get the highest data_type in the possible dependencies
+        data_types = sorted(
+            self.context.get_dependencies(data_type) & (possible_dependencies),
```
Comment (rsriya) [inline outsource/config.py:135]: Why is self.context.get_dependencies(data_type) not sufficient? Why do we need & operation?

File: outsource/meta.py (hunk 3 above)
Comment (dachengx) [reply outsource/meta.py:3]: `records` are not always saved by straxen: https://github.com/XENONnT/straxen/blob/ff7c787e7769cfc05192c96a3b19c51eaedc7df1/straxen/plugins/records/records.py#L44. So I disallow records being in include_data_types. The processing of `records` will only be a temporary step of the processing of `peaklets`.

File: outsource/config.py (hunk 4 above)
Comment (dachengx) [reply outsource/config.py:90]: Here the `RunConfig.depends_on` returns data_type in `PER_CHUNK_DATA_TYPES` or `raw_records*`. So the `&` here limits the range of searching. The `RunConfig.depends_on` is not the same as `depends_on` in strax, but is a guide on which important data_type we should look at when checking the available RSE. You can search for the usage of `RunConfig.depends_on` to look at the details.

File: outsource/meta.py (hunk 2 above)
Comment (dachengx) [reply outsource/meta.py:16]: In the new logic, we will save `peak_basics` and `event_info` when processing `event_info_double`. And we do not allow `peak_basics` and `event_info` to be in `include_data_types` in `XENON_CONFIG`.

I will check with Lanqing about whether we saved `event_area_per_channel `.

**Pull request merged**